monai/utils/tf32.py

```diff
@@ -0,0 +1,79 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import functools
+import os
+import warnings
+
+from monai.utils.decorators import reset_torch_cuda_after_run
+
+__all__ = ["has_ampere_or_later", "detect_default_tf32"]
+
+
+@functools.lru_cache(None)
+@reset_torch_cuda_after_run
+def has_ampere_or_later() -> bool:
+    """
+    Check if there is any Ampere and later GPU.
+    """
+    import torch
+
+    from monai.utils.module import version_geq
+
+    if not version_geq(f"{torch.version.cuda}", "11.0"):
+        return False
+    for i in range(torch.cuda.device_count()):
+        major, _ = torch.cuda.get_device_capability(i)
+        if major >= 8:  # Ampere and later
+            return True
```

Comment:
Two alternatives that do not use `torch.cuda` for this:
1.
`=compute_cap` flag is available after 11.6(not sure)
```py
try:
    capabilities_str = subprocess.run(
        ["nvidia-smi", "--query-gpu=compute_cap", "--format=csv,noheader"],
        check=True,
        capture_output=True,
        text=True
    ).stdout
    capabilities_list = capabilities_str.split('\n')
    if any([float(c) >= 8 for c in capabilities_list if c]):
        return True
except BaseException:
    pass
```
2.
[pyNVML](https://pypi.org/project/nvidia-ml-py/) 36KB
```py
from pynvml import nvmlInit, nvmlShutdown, nvmlDeviceGetCount, nvmlDeviceGetHandleByIndex, nvmlDeviceGetCudaComputeCapability
try:
    nvmlInit()
    for i in range(nvmlDeviceGetCount()):
        handle = nvmlDeviceGetHandleByIndex(i)
        major, _ = nvmlDeviceGetCudaComputeCapability(handle)
        if major >= 8:
            return True
except BaseException:
    pass
finally:
    nvmlShutdown()
```

Reply:
I'm trying to get back up to speed on MONAI stuff, but here I would avoid external programs since this can be slow and will break if the program isn't found through PATH, eg. under Windows in some cases. The pyNVML solution would be better and would be minor dependency to add to MONAI.

Reply:
> I'm trying to get back up to speed on MONAI stuff, but here I would avoid external programs since this can be slow and will break if the program isn't found through PATH, eg. under Windows in some cases. The pyNVML solution would be better and would be minor dependency to add to MONAI.

Thanks for looking into this. Could you also check the `run_command()` of `_version.py`? It uses `subprocess.Popen`, and per my understanding of https://stackoverflow.com/a/39187984/10973766, it will not block the `import monai` process.

Reply:
However, the PATH on windows of alt 1. is indeed unreliable.

Reply:
> > I'm trying to get back up to speed on MONAI stuff, but here I would avoid external programs since this can be slow and will break if the program isn't found through PATH, eg. under Windows in some cases. The pyNVML solution would be better and would be minor dependency to add to MONAI.
> 
> Thanks for looking into this. Could you also check the `run_command()` of `_version.py`? It uses `subprocess.Popen`, and per my understanding of https://stackoverflow.com/a/39187984/10973766, it will not block the `import monai` process.

Yeah I had seen before how this file works and it's a similar issue. [Here](https://github.com/Project-MONAI/MONAI/blob/65cf5fe1234d0d01cc52c9ab48395a3efbad6f45/monai/_version.py#L232) Popen is used to call git and parse out information about version and where the HEAD currently is. This could be parsed from data in the .git directory directly, which is what git does, but would be more complex and possibly error prone if the formats of these files change. It's arguably more portable to rely on `git/git.cmd/git.exe` to do the parsing for you and simple to use but still can break if the executable isn't found. On a regular desktop environment the main walltime cost when importing MONAI is still initializing CUDA in Pytorch so that's not a concern but something I'd prefer to redo sometime.